### PR TITLE
Adapted for patch_radius with radii differing xyz direction. 

### PR DESCRIPTION
**Pull request by MNagtegaal**
… behaviour is maintained.

### CODE REVIEW

**Comment by ShreyasFadnavis** [comment]
Hi @MNagtegaal ! Thank you for this PR :+1: can you also add a small unit test for when this adaptation of the patch radius will fail/ not be reasonable? 

Perhaps also a `warning` for when the matrix decomposition will become ill-posed and the user should try a different patch radius?

Otherwise this PR looks good!!

File: dipy/denoise/localpca.py [hunk 1]
```diff
@@ -139,11 +139,16 @@ def genpca(arr, sigma=None, mask=None, patch_radius=2, pca_method='eig',
...
     patch_size = 2 * patch_radius + 1
 
-    if patch_size ** 3 < arr.shape[-1]:
+    if np.prod(patch_size) < arr.shape[-1]:
         e_s = "You asked for PCA denoising with a "
         e_s += "patch_radius of {0} ".format(patch_radius)
+        e_s += "with total patch size of {0}".format(np.prod(patch_size))
         e_s += "for data with {0} directions. ".format(arr.shape[-1])
         e_s += "This would result in an ill-conditioned PCA matrix. "
         e_s += "Please increase the patch_radius."
```
Comment (MNagtegaal) [inline dipy/denoise/localpca.py]: Is this the kind of unit test you were thinking of? @ShreyasFadnavis

**Comment by MNagtegaal** [comment]
Is the above commented change what you were thinking of?

File: dipy/denoise/localpca.py (hunk 1 above)
Comment (ShreyasFadnavis) [reply dipy/denoise/localpca.py]: Yep! Thanks!

Comment (ShreyasFadnavis) [reply dipy/denoise/localpca.py]: You will also need to adapt the tests like:
https://github.com/nipy/dipy/blob/9f3dc9b58e3b6b7390c08968c07822a28c0fe1f3/dipy/denoise/tests/test_lpca.py#L242

With the new type of patch radius that you have proposed... does this make sense?

Comment (ShreyasFadnavis) [reply dipy/denoise/localpca.py]: > You will also need to adapt the tests like:
> https://github.com/nipy/dipy/blob/9f3dc9b58e3b6b7390c08968c07822a28c0fe1f3/dipy/denoise/tests/test_lpca.py#L242
> 
> With the new type of patch radius that you have proposed... does this make sense?

See the whole file.. that will make Travis happy👌🏻

**Comment by MNagtegaal** [comment]
@ShreyasFadnavis I added tests for too small radii and patch_radius of the wrong size. 
I have to say that this is my first encounter with these testing procedures, so I might overlook some things.

**Comment by ShreyasFadnavis** [comment]
Hi @MNagtegaal almost there.. there are some more functions that you need to adapt in the test file. You can see where the changes need to be made from the log of the Travis CI in the link below:
https://travis-ci.org/nipy/dipy/jobs/633782976?utm_medium=notification&utm_source=github_status

It should give you the function names of the test functions that need to be changed...

**Comment by MNagtegaal** [comment]
I see that there was still some mistake. Could I have ran these tests before making the pull request?

**Comment by ShreyasFadnavis** [comment]
> I see that there was still some mistake. Could I have ran these tests before making the pull request?

Yes! You can just execute the `test_lpca.py` file and it should show you where the tests are failing. 
- In DIPY however, we use `pytest` : https://docs.pytest.org/en/latest/ to setup the test suite.

Further details can be found in the [CONTRIBUTING.md](https://github.com/nipy/dipy/blob/740a27cbae71baaf19c66c3151801b988da04428/CONTRIBUTING.md) file!

**Comment by skoudoro** [comment]
Thank you for this cool update @MNagtegaal! Nice Job! and thank you for this review @ShreyasFadnavis!

Overall, Looks good to me! I will try to play with this feature today and merge it by the end of the day if no one has any other comments.

**Comment by ShreyasFadnavis** [comment]
@MNagtegaal @skoudoro just realized: We may need to update the `MP-PCA example`. 

But can be on a separate PR I guess 👍 Just a couple of lines need to be added I guess...

**Comment by Garyfallidis** [comment]
@MNagtegaal are sure everything is okay close to the boundaries of the image?

**Comment by MNagtegaal** [comment]
There are no changes how the boundaries are dealt with, so that should be ok.

**Comment by ShreyasFadnavis** [comment]
> There are no changes how the boundaries are dealt with, so that should be ok.

+1 -- @Garyfallidis : I don't think boundaries will be affected by this!

**Comment by skoudoro** [comment]
Thank you a lot for your contribution @MNagtegaal! merging!

**Pull request merged**